### PR TITLE
Simplify context store read check

### PR DESCRIPTION
**Pull request by mcculls**
The partial JVM simulation made this code more complex than it needed to be for the purposes of checking the last two constants loaded before the method invocation. Also replaced use of reflected method information with constants - this avoids repeated string transformations while still catching the necessary rewriting.

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java [hunk 1]
```diff
@@ -107,24 +110,20 @@ public void visitMethodInsn(
...
               `InstrumentationContext.get(K.class, V.class)`. If it does the inside of this if rewrites it to call
               dynamically injected context store implementation instead.
                */
-              if ((insnStack[0] == Opcodes.INVOKESTATIC
-                      && insnStack[1] == Opcodes.LDC
-                      && insnStack[2] == Opcodes.LDC)
-                  && (stack[0] instanceof Type && stack[1] instanceof Type)) {
-                final String contextClassName = ((Type) stack[0]).getClassName();
-                final String keyClassName = ((Type) stack[1]).getClassName();
+              if (constant_1 instanceof Type && constant_2 instanceof Type) {
```
Comment (richardstartin) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java:124]: Sorry to bikeshed but can we not do snake case?

Comment (mcculls) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java:124]: sure, no worries

**Review by richardstartin — APPROVED** [review]

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java]: c7dcd0d45

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java [hunk 2]
```diff
@@ -107,24 +110,20 @@ public void visitMethodInsn(
...
+                && INSTRUMENTATION_CONTEXT_CLASS.equals(owner)
+                && GET_METHOD.equals(name)
+                && GET_METHOD_DESCRIPTOR.equals(descriptor)) {
               log.debug("Found context-store access in {}", instrumenterClassName);
               /*
               The idea here is that the rest if this method visitor collects last three instructions in `insnStack`
               variable. Once we get here we check if those last three instructions constitute call that looks like
               `InstrumentationContext.get(K.class, V.class)`. If it does the inside of this if rewrites it to call
               dynamically injected context store implementation instead.
                */
```
Comment (bantonsson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java:123]: This comment can use an update.

**Review by bantonsson — APPROVED** [review]
So the code is a bit more relaxed now, and the instructions don't need to be directly after each other, but I can't see any issues with that.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java (hunk 2 above)
Comment (mcculls) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/ContextStoreReadsRewritingVisitor.java]: 012b95b8a

**Comment by bantonsson** [comment]
Just to be clear, the new code can do the wrong thing if we have code that adds other things onto the stack in between the constant loads and the actual call, but since this is for transforming code that is under our control, I think it is fine to be more relaxed.

**Comment by mcculls** [comment]
Yes this assumes the bytecode follows the typical convention of loading arguments just before the call, which I think is reasonable.

Note the old code wasn't a perfect simulation of the JVM - its stack was very limited, nothing ever got popped, and not all opcodes were covered. So it also had the potential for mis-identifying the constants for non-typical bytecode, just a smaller subset.

**Pull request merged**